lib/acl.c

```diff
@@ -43,17 +43,36 @@ enum {
 	NUM_INPUTS_IPV6,
 };
 
-/* Callback function for when there's no classification match. */
 static inline int
-drop_ipv6_acl_pkts(struct rte_mbuf **pkts, unsigned int num_pkts,
-	__attribute__((unused)) struct gatekeeper_if *iface)
+drop_ipv6_acl_pkts(struct rte_mbuf **pkts, int num_pkts)
 {
-	unsigned int i;
+	int i;
 	for (i = 0; i < num_pkts; i++)
 		rte_pktmbuf_free(pkts[i]);
 	return 0;
 }
 
+/* Callback function for when there's no classification match. */
+static inline int
+match_ipv6_acl_pkts(struct rte_mbuf **pkts, unsigned int num_pkts,
+	struct gatekeeper_if *iface)
+{
+	unsigned int i;
+	for (i = 0; i < num_pkts; i++) {
+		unsigned int j;
+		/* Start from the first one. */
+		for (j = 1; j < iface->acl_func_count; j++) {
+			int ret = iface->ext_funcs[j](pkts[i], iface);
+			if (ret == 0)
+				break;
+		}
+
+		if (j == iface->acl_func_count)
+			rte_pktmbuf_free(pkts[i]);
```

Comment:
Functions like `submit_bgp()` and `submit_nd()` are responsible for freeing packets if something goes wrong, so this `rte_pktmbuf_free()` should really only be called if the loop completes. In other words, the `break` inside the for loop is a bug, since we could be calling a function like `match_bgp()` which fails when running `submit_bgp()`, which would have already freed the packets.